sc-memory/sc-memory/sc_agent.tpp

```diff
@@ -0,0 +1,632 @@
+
+    SC_LOG_INFO(
+        "Unsubscribe agent `" << agentName << "` from event `" << eventClassName << "("
+                              << subscriptionElementAddr.Hash() << ")`.");
+
+    delete it->second;
+    subscriptionsMap.erase(subscriptionElementAddr);
+  }
+}
+
+template <class TScEvent, class TScContext>
+template <class TScAgent>
+std::function<void(TScEvent const &)> ScAgent<TScEvent, TScContext>::GetCallback(
+    ScAddr const & agentImplementationAddr) noexcept
+{
+  static_assert(std::is_base_of<ScAgent, TScAgent>::value, "TScAgent type must be derived from ScAgent type.");
+
+  static_assert(
+      HasNoMoreThanOneOverride<TScAgent>::InitiationConditionMethod::value,
+      "TScAgent must have no more than one override method from methods: `GetInitiationCondition`, "
+      "`GetInitiationConditionTemplate` "
+      "and `CheckInitiationCondition`.");
+
+  static_assert(
+      HasNoMoreThanOneOverride<TScAgent>::ResultConditionMethod::value,
+      "TScAgent must have no more than one override method from methods: `GetResultCondition`, "
+      "`GetResultConditionTemplate` and "
+      "`CheckResultCondition`.");
+
+  static_assert(
+      HasOneOverride<TScAgent>::DoProgramMethod::value, "TScAgent must have one override `DoProgram` method.");
+
+  return [agentImplementationAddr](TScEvent const & event) -> void
+  {
+    auto const & ResolveAction = [](TScEvent const & event, ScAgent & agent) -> ScAction
+    {
+      auto [subscriptionElementAddr, _, otherElementAddr] = event.GetTriple();
+      if (subscriptionElementAddr == ScKeynodes::action_initiated)
+      {
+        ScAddr const & actionAddr = otherElementAddr;
+        return agent.m_context.ConvertToAction(actionAddr);
+      }
+
+      return agent.m_context.CreateAction(agent.GetActionClass()).Initiate();
+    };
+
+    TScAgent agent;
+    std::string const & agentName = agent.GetName();
+    SC_LOG_INFO("Agent `" << agentName << "` started");
+
+    agent.SetInitiator(event.GetUser());
+    agent.SetImplementation(agentImplementationAddr);
+    if (agent.IsActionClassDeactivated())
+    {
+      SC_LOG_WARNING(
+          "Agent `" << agentName << "` was finished because actions with class `" << agent.GetActionClass().Hash()
+                    << "` are deactivated.");
+      return;
+    }
+
+    ScAction action = ResolveAction(event, agent);
+    if (!agent.template ValidateInitiationCondition<TScAgent>(event))
+    {
+      SC_LOG_WARNING(
+          "Agent `" << agentName << "` was finished because its initiation condition was checked unsuccessfully.");
+      return;
+    }
+
+    ScResult result;
+
+    try
+    {
+      if constexpr (HasOverride<TScAgent>::DoProgramWithEventArgument::value)
+        result = agent.DoProgram(event, action);
+      else
+        result = agent.DoProgram(action);
+    }
+    catch (utils::ScException const & exception)
+    {
+      action.FinishWithError();
```

Comment:
exception may be thrown here(if someone erases action node and then tries to finish action)

Reply:
yes it does, but exception can be caught and agent can log that it couldn't finish action with error because another error happened

Reply:
Fixed